### PR TITLE
Add support for custom emoji template

### PR DESCRIPTION
**Pull request by nodegin**
Hi,

While using twemoji, we wanted to let user select parsed emojis like normal text, however, we are unable to achieve this by styling the img tag alone because we need:

- `pointer-event: none` to let user select image like normal text
- `cursor: text` to change cursor for UX

However `pointer-event: none` clashes with `cursor: text` on the same element, we needed to create a container that wraps the emoji element in order to achieve what we want.

We are creating this PR hoping twemoji could implement built in support for such functionality.

Usage with the new template option:

```es
twemoji.parse(htmlText, {
  template: (emoji) => {
    return `
      <div class="emoji-container">
        ${emoji}
      </div>
    `
  },
})
```

### CODE REVIEW

**Comment by WebReflection** [comment]
Will have a look next week (now abroad) thanks

**Comment by WebReflection** [comment]
there is one bit I don't like too much:

```js
// template function for custom emoji wrapping
```

a *tempalte* for *wrapping* doesn't sound congruent with the names chosen to add *wrapping* functionality.

The current emoji template is already the image with a custom `src` or a `className`.

That means that you can inject the parsed string and right away do the following:
```js
// DOM4 + ES6
twemoji.parse(container);
container.querySelectorAll('img.emoji').forEach(img => {
  const div = img.ownerDocument.createElement('div');
  img.replaceWith(div);
  div.append(img);
})
```

If you are generating emoji on the server you can obtain similar results through a RegExp that catches `//g` and replace with surrounding div.

### As Summary
I'm not sure this functionality is really needed in core but we might just bring it in as long as:

  * there are tests that shows this functionality actually works with both strings and DOM nodes. Current proposal assumes people uses strings when actually most developers use DOM and this will break.
  * the changes are done to the correct file which is the generate file, not the lib itself.
  * the name is not actually `template` because we are talking about a *wrap* and it could be either for strings or DOM nodes (the logic should be compatible with DOM too, the `ret = template(ret)` is not going to work here)

**Comment by nodegin** [comment]
Hi @WebReflection,

Thanks for the feedback and suggestion, much appreciated.

We are already using a similar approach to your suggestion for production. We ran some benchmarks and the results shows that in order to achieve what we want, DOM parsing is up to about 4 times slower than our proposal.

We believe that this new option provides a simpler way for users who wants to apply container elements to the emoji or modify the emoji element, with the benefits of better performance and less code required.

We agree that 'template' might not be the best name, we are open to any name suggestions.

**Comment by WebReflection** [comment]
beside the name, the logic is not compatible with current API because it accepts **both** strings and DOM nodes.

The `template-but-not-really` method should not just return a string but a DOM node when DOM is used instead of strings, which is the most secure way to use twemoji, hence the preferred/suggested one.

**Summary**
the follwing:
```js
twemoji.parse(htmlText, {
  template: (emoji) => {
    return `
      <div class="emoji-container">
        ${emoji}
      </div>
    `
  },
})
```
should work also via DOM input
```js
twemoji.parse(htmlText, {
  template: (emoji) => {
    var div = emoji.ownerDocument.craeteElement('div');
    div.className = 'emoji-container';
    emoji.parentNode.replaceNode(div, emoji);
    div.appendChild(emoji);
    return div;
  },
})
```

or something similar.

Long story short, this PR is incomplete and incompatible with current API as it is.

**Comment by WebReflection** [comment]
P.S. we've tested DOM vs strings too and if you pass all nodes at once you shouldn't have that bad performance. The key is to pass incremetally all new nodes instead of hwole document and you could use a document fragment for that. Anyway, let's try to adress outstanding issues first and see if we can put this in
